llmc/bits.cuh

```diff
@@ -0,0 +1,97 @@
+#if defined(_MSC_VER) && !defined(__clang__) // MSVC
+    __assume(false);
+#else // GCC, Clang
+    __builtin_unreachable();
+#endif
+}
+
+// ----------------------------------------------------------------------------
+//  bit-fiddling to reconstruct master weights from BF16 and missing bits
+// handling bf16 is _almost_ just storing the 16 bits of significant that get
+// truncated when going from float -> bf16; except we do stochastic rounding.
+// if we end up rounding towards zero, we could just keep the bits, but if we
+// round away from zero, there is a chance that the other bits of the bf16 no
+// longer correspond to the bits in the original float. So we have to reserve
+// one bit to remember whether we rounded up or down, for an effective master
+// weight precision of fp31. That should still be more than sufficient.
+
+// Result of splitting a float into a stochastically-rounded bfloat16 and
+// additional reconstruction bits
+struct SplitFloatResult {
+    nv_bfloat16 b_float;
+    unsigned short bits;
+};
+
+// UB-free bit-level conversions. A C++17 version for C++20 std::bit_cast
+template<class T, class S>
+__host__ __device__ T bit_cast(S v) {
+    T dest;
+    static_assert(sizeof(v) == sizeof(dest), "Size mismatch.");
+    memcpy(&dest, &v, sizeof(v));
+    return dest;
+}
+
+// Splits a float into a bfloat16 and the remaining significant bits
+__host__ __device__ SplitFloatResult split_float(float value, unsigned short threshold) {
+    unsigned int float_bits = bit_cast<unsigned int>(value);
+    // IEEE 754: float: S E(8) M (23)    bfloat: same, but significant 23-16 = 7 bits
+    // ideally, we'd just store the cut-off 16 bits, but that doesn't work if rounding
+    // is involved.
+    unsigned int rounded_bits = float_bits & 0x0000FFFFu;
+    if(rounded_bits > threshold) {
+        SplitFloatResult result;
+        result.b_float = __float2bfloat16_rn(bit_cast<float>(float_bits | 0xFFFFu));
+        result.bits = rounded_bits & (~1u) | 1u;
+        return result;
+    } else {
+        // truncation is easy
+        SplitFloatResult result;
+        result.b_float = bit_cast<__nv_bfloat16>((unsigned short)(float_bits >> 16u));
+        result.bits = rounded_bits & (~1u);
+        return result;
+    }
+}
+
+// Reassembles a float from the bfloat16 part and the missing mantissa
+__host__ __device__ float assemble_float(nv_bfloat16 b_float, unsigned short bits) {
+    constexpr const unsigned short BF16_SIGN_MASK        = 0b1'00000000'0000000u;
+    constexpr const unsigned short BF16_EXPONENT_MASK    = 0b0'11111111'0000000u;
+    constexpr const unsigned short BF16_SIGNIFICANT_MASK = 0b0'00000000'1111111u;
+    unsigned short bf = bit_cast<unsigned short>(b_float);
+    if(bits & 1u) {
+        // if we rounded away from zero, we need to undo these changes.
+        // first, check if the significant (7 bits) of bf16 is zero
+        const unsigned short significant = bf & BF16_SIGNIFICANT_MASK;
```

Comment:
nit: i think people usually refer to this one as `significand`

Reply:
right. Mantissa might actually be the correct term here. From wikipedia:
> To understand both terms, notice that in binary, 1 + mantissa ≈ significand, and the correspondence is exact when storing a power of two. This fact allows for a fast approximation of the base-2 logarithm, leading to algorithms e.g. for computing the [fast square-root](https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Approximations_that_depend_on_the_floating_point_representation) and [fast inverse-square-root](https://en.wikipedia.org/wiki/Fast_inverse_square_root). The implicit leading 1 is nothing but the hidden bit in IEEE 754 floating point, and the bitfield storing the remainder is thus the mantissa.

Reply:
good point! mantissa is definitely the right term then

Reply:
changed to mantissa